include/wil/filesystem.h

```diff
@@ -624,10 +624,15 @@ namespace wil
 
                     if (m_folderHandle)
                     {
-                        CancelIoEx(m_folderHandle.get(), &m_overlapped);
-
-                        DWORD bytesTransferredIgnored = 0;
-                        GetOverlappedResult(m_folderHandle.get(), &m_overlapped, &bytesTransferredIgnored, TRUE);
+                       BOOL cancelRes = CancelIoEx(m_folderHandle.get(), &m_overlapped);
+
+                       // no pending operation to cancel. Maybe StartIo returned 
+                       // an error?
+                       if (!(cancelRes == FALSE && ::GetLastError() == ERROR_NOT_FOUND)) 

```

Comment:
parens around sub-expressions. 
here is a re-write using DeMorgan, but this seems strange to me. is this right?
`if (cancelRes || (::GetLastError() != ERROR_NOT_FOUND))`

Reply:
Yes, it looks odd to me too (hard to parse) 
Also I am not sure when we should call `GetOverlappedResult`:
- If `CancelIoEx` returned `TRUE` -- we should always call it.
- If `CancelIoEx` returned false:
-- do we want to call it when `GetLastError() != ERROR_NOT_FOUND` ? 
-- should we never call it? (maybe there is another error that could cause the call to hang).

Reply:
I don't know either so would need to consult with experts. maybe I can get some of their attention. do you have a repro case we can point to? ideally in the form of a test.

Reply:
From an expert... (thank you!)

> If I’m reading it right, CancelIoEx can fail because:
Kernelbase passed ntoskrnl bad memory (shouldn’t happen);
The handle is invalid (shouldn’t happen);
The request is not outstanding so couldn’t be cancelled (what this fix is about);
The request is outstanding but was already cancelled by another thread so this thread won’t cancel it.  I think wil is already protecting itself against that by the m_cancelLock above this call; this is the challenging case because if it happened it’s still important to wait for the request to complete, even if this thread didn’t initiate the cancel.  So my 2c would be that this change looks fine, although it shouldn’t be blindly copy-pasted to an environment without that cancel lock.

Reply:
So, how should we go forward with this change?